snuba/query/allocation_policies/__init__.py

```diff
@@ -0,0 +1,207 @@
+        >>>     QueryResultOrError(result=result)
+        >>> )
+
+    The allocation policy base class has two public methods:
+        * get_quota_allowance
+        * update_quota_balance
+
+    These functions can be used to modify the behaviour of ALL allocation policies, use with care.
+
+    **GOTCHAS**
+    -----------
+
+    * At time of writing (29-03-2023), not all allocation policy decisions are made in the allocation policy,
+        rate limiters are still applied in the query pipeline, those should be moved into an allocation policy as they
+        are also policy decisions
+    * get_quota_allowance will throw an AllocationPolicyViolation if _get_quota_allowance().can_run is false.
+        this is to keep with the pattern in `db_query.py` which communicates error states with exceptions. There is no other
+        reason. For more information see snuba.web.db_query.db_query
+    * Every allocation policy takes a `storage_set_key` in its init. The storage_set_key is like a pseudo-tenant. In different
+        environments, storages may be co-located on the same cluster. To facilitate resource sharing, every allocation policy
+        knows which storage_set_key it is serving. This is currently not used
+    """
+
+    def __init__(
+        self,
+        storage_set_key: StorageSetKey,
+        required_tenant_types: list[str],
+        **kwargs: str,
+    ) -> None:
+        self._required_tenant_types = set(required_tenant_types)
+        self._storage_set_key = storage_set_key
+
+    @classmethod
+    def config_key(cls) -> str:
+        return cls.__name__
+
+    @classmethod
+    def get_from_name(cls, name: str) -> "AllocationPolicy":
+        return cast("AllocationPolicy", cls.class_from_name(name))
+
+    def __eq__(self, other: Any) -> bool:
+        """There should not be a need to compare these except that
+        AllocationPolicies are attached to the Table a query is executed against.
+        In order to allow that comparison, this function is implemented here.
+        """
+        return (
+            bool(self.__class__ == other.__class__)
+            and self._storage_set_key == other._storage_set_key
+            and self._required_tenant_types == other._required_tenant_types
+        )
+
+    @classmethod
+    def from_kwargs(cls, **kwargs: str) -> "AllocationPolicy":
+        required_tenant_types = kwargs.pop("required_tenant_types", None)
+        storage_set_key = kwargs.pop("storage_set_key", None)
+        assert isinstance(
+            required_tenant_types, list
+        ), "required_tenant_types must be a list of strings"
+        assert isinstance(storage_set_key, str)
+        return cls(
+            required_tenant_types=required_tenant_types,
+            storage_set_key=StorageSetKey(storage_set_key),
+            **kwargs,
+        )
+
+    def get_quota_allowance(self, tenant_ids: dict[str, str | int]) -> QuotaAllowance:
```

Comment:
How often is this method called? Is the expectation that it is very fast? I'm wondering if there would be a huge perf impact if say someone had a complicated statistical algo here.